### PR TITLE
WIP: Add libgcc split

### PR DESCRIPTION
**Pull request by jakirkham**
Here we do something that has long been discussed and people have generally been amenable to. This breaks up the `libgcc` library into a bunch of smaller library packages.

The hope is that people can be more particular with what they pull in dependency wise. For instance, if someone is just using GFortran, they should only need to add the `libgfortran` package. That way `libstdc++` or something else that is unneeded is not included. Same story with using GNU OpenMP, one need only include `libgomp`. However, people that are using `libstdc++` may still present problems. Not sure what more to do about that ATM.

I have decided to version these based on the `gcc` compiler they were pulled from. IMHO this make simple to understand where they came from and makes it easier to stay on top of updating them. Also IMHO `defaults` should follow suit.

cc @pelson @ocefpaf @msarahan @gillins @patricksnape @jjhelmus

### CODE REVIEW

File: recipes/libgcc/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    - libquadmath {{ version }}
+    - libstdc++ {{ version }}
+
+build:
+  number: 0
+  skip: true                    # [win]
+
+test:
+  commands:
+    - true
```
Comment (jakirkham) [inline recipes/libgcc/meta.yaml]: Thinking that we could test that all the other libraries are present, but that also creates an annoying maintenance burden. Feedback welcome.

File: recipes/libgfortran/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+package:
+  name: libgfortran
+  version: {{ version }}
+
+requirements:
+  build:
+    - gcc {{ version }}
+
+  run:
+    - libquadmath {{ version }}
```
Comment (jakirkham) [inline recipes/libgfortran/meta.yaml]: This is true on both platforms. This way we can just add it outright.

File: recipes/libgomp/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+package:
+  name: libgomp
+  version: {{ version }}
+
+requirements:
+  build:
+    - gcc {{ version }}
+  run:
+    - libgcc_s {{ version }}    # [osx]
```
Comment (jakirkham) [inline recipes/libgomp/meta.yaml]: Appears Linux does not have this dependency. Only OS X has it. Bit strange, but that's what `ldd`/`otool -L` tell us.

File: recipes/libstdc++/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+package:
+  name: libstdc++
+  version: {{ version }}
+
+requirements:
+  build:
+    - gcc {{ version }}
+  run:
+    - libgcc_s {{ version }}
```
Comment (jakirkham) [inline recipes/libstdc++/meta.yaml]: This is the case on both platforms.

**Comment by jakirkham** [comment]
Also, please let me know if you want to be a maintainer. I'm not sure that I'm going to stay on top of these alone.

**Comment by jakirkham** [comment]
Understandably the linter is unhappy. I just copied the conda-recipes and broke it into pieces. So, looks like the order is wrong.

File: recipes/libgcc/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
+{% set version = "4.8.5" %}
+
+package:
+  name: libgcc
+  version: {{ version }}
+
+build:
+  number: 0
```
Comment (jakirkham) [inline recipes/libgcc/meta.yaml]: Sadly we are probably going to need to bump this a bit.

**Comment by ocefpaf** [comment]
I am seriously concerned about the maintainability and stability of all these packages. I would love to hear @msarahan opinion here as continuum has been doing this for a while and has a lot of experience to share.

**Comment by msarahan** [comment]
We did some thinking on this before, but unfortunately the issue was in a private repo.  I moved it to a public one:

https://github.com/ContinuumIO/anaconda-issues/issues/843

TL;DR: please do not version libraries with the gcc version.  Instead, use metapackages to track ABI version.  This will be much more straightforward.

**Comment by jakirkham** [comment]
Before we get into the versioning discussion, I'd like to ask some feedback on a simple issue of getting this to build on the docker container. The `gcc` package won't install itself because of the `crtxxx.o` files. In the past, I have install `gcc` from `yum` to solve this problem. This works, but are there other simpler solutions.

**Comment by jakirkham** [comment]
Sorry, maybe it is `isl` version issue.

**Comment by jakirkham** [comment]
Yep, it is an `isl` version issue. Not sure how we want to handle this. I have forced our `cloog` workaround for now. Though I really don't love that such a workaround is required.

**Comment by jakirkham** [comment]
See this related issue ( https://github.com/ContinuumIO/anaconda-issues/issues/833 ) for hot-fixing `gcc`'s `isl` pinning. Have bumped the issue to see if we can do something about this.

**Comment by jakirkham** [comment]
See this [link]( https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html ) as an example.

**Comment by pelson** [comment]
> I am seriously concerned about the maintainability and stability of all these packages.

I'm in agreeance simply because I don't think we have a need for all of these splits at present. AFAIK (and I have been out of the loop, so please do correct me if I'm wrong) the only separated package which we really need is `libgfortran`? Can we *just* do that one for now until we have a concrete usecase for the others?

I'm happy to be named as a maintainer of libgfortran.

**Comment by msarahan** [comment]
C++ is confusing, and may require an extra metapackage to capture the CXXABI.

In general, you have the SO version of a library.  The version of the library package that actually contains files should have this SO version.  For libstdc++, this captures GLIBCXX.  For libgfortran, SO version is 3 right now, but internal symbol version is something else (6 maybe?).  There needs to be a metapackage representing that symbol version, and the library package needs to be associated with it somehow.  At this point, this information is all just being stored.  In the future, we are working towards recording what versions of tools (and ABIs) were used, and use that info to obtain constraints on what packages are compatible.  Using different compiler packages may introduce these constraints into built package dependency metadata.

**Comment by jakirkham** [comment]
Well, there are things that use `libgomp` independent of anything else in `libgcc`. If we leave them dependent on `libgcc`, we will simply run into the stated problem with a different use case. Keep in mind the only mode of parallelism provided by `cython` ATM is OpenMP. So, this is fairly common.

The value of splitting `libquadmath` is that someone could build code with `float128` support, but not necessarily need anything else (think C). This gives them that option. It is also a requirement of `libgfortran`. So, it made sense to break this out.

`libgcc_s` appears to only be needed by `libgomp` on OS X (though this does seem odd 😕). It is also used by `libstdc++`. Instead of including that logic twice, it seemed smarter to split it out into its own self-contained package.

At this point, we only have `libstdc++` and I felt it should be addressed in the same way.

More generally, based on the working hypothesis, pulling down the full `libgcc` is becoming problematic. If this is in fact the issue, it makes sense IMHO to break this down to the level where we are pulling in only what we need and not portions that break a users system. However, for compatibility reasons (and probably user expectations) the `libgcc` package needs to remain in some form. So, we have a metapackage that has all of its content broken out.

This all being said, I'm willing to break this down into smaller more solvable problems. Perhaps `libgfortran`/`libquadmath` would be a good start.

However, I think this level of granularity (not the versioning nor the package structure as is) is the direction we want to go ultimately. Based on a consensus of several people at Continuum, it seems that they also agree with this level of granularity. Please see the linked issue ( https://github.com/ContinuumIO/anaconda-issues/issues/843 ) that @msarahan has provided.

**Comment by pelson** [comment]
Let's try no to be too general at this moment and solve `libgfortran` alone. It is perfectly reasonable to simply pull `libquadmath`'s `.so` into `libgfortran` until we have a need for `float128` support elsewhere IMO (I fully accept that the solution isn't general, and that we *will* want to iterate on it in the future).

In this matter I'm quite content for Continuum to lead on the gcc split - if they come up with a proven solution superior then I'm completely happy to adopt it.

**Comment by jakirkham** [comment]
SGTM. Given @pelson's suggestion, how would you recommend we do versioning for `libgfortran`, @msarahan?

**Comment by ocefpaf** [comment]
> In this matter I'm quite content for Continuum to lead on the gcc split - if they come up with a proven solution superior then I'm completely happy to adopt it.

It is important for us to work with Continuum on this to avoid mismatch between `conda-forge` and `defaults`. What is the difference between what we already have in `defaults`, which has a `libgfortran`, and the proposal here? Why does `libgfortran` only did not work for `openblas`?

**Comment by jakirkham** [comment]
So, we could not use `libgfortran` as there was none for Mac.

There were also some version issues with the `libgfortran` package on Linux that I did not completely follow as I don't know how this package was created. Am still a bit worried about whether it is causing problems or not currently. Hoping someone else can fill in the blanks here.

**Comment by jakirkham** [comment]
I have broken out the existing stuff for `libgfortran`/`libquadmath` as one package into this PR ( https://github.com/conda-forge/staged-recipes/pull/861 ). Let's please discuss that point over there. Thanks.

**Comment by msarahan** [comment]
The version problems that bit people were that many early versions of numpy (1.9.x) did not pin ligfortran at all.  So libgfortran update from 1.x to 3.x broke those people.

Next, if anyone used a gfortran newer than ours to build something, then their library would go looking for symbols that weren't in libgfortran (ABI mismatch).

Starting with libgfortran is a good start.  Please make libgcc depend (run) on:

libgfortran 3.*
libgfortran-abi >=1.6  (I think - please verify this, as below)

```
[dev@98227d75440e site-packages]$ strings /usr/local/lib64/libgfortran.so | grep GFORTRAN
GFORTRAN_1.0
GFORTRAN_1.1
GFORTRAN_1.2
GFORTRAN_1.3
GFORTRAN_1.4
GFORTRAN_1.5
GFORTRAN_1.6
```

**Comment by ocefpaf** [comment]
> So, we could not use `libgfortran` as there was none for Mac.

Good. So it was not the model of having a `libgfortran`. Meaning that the solutions used for Linux right now would also work on OS X and there is no need for the split.

> I have broken out the existing stuff for `libgfortran`/`libquadmath` as one package

:+1: as that is closer to what is done in the default channel.

**Comment by jakirkham** [comment]
Going to close this out as there will be no further work on this at this time. Still think this is very important and this general strategy should be pursued. Just not sure that it should be done at this time.
